taichi/transforms/make_thread_local.cpp

```diff
@@ -0,0 +1,150 @@
+bool is_atomic_op_linear(AtomicOpType op_type) {
+  return op_type == AtomicOpType::add || op_type == AtomicOpType ::sub;
+}
+
+void make_thread_local_offload(OffloadedStmt *offload) {
+  // TODO: deal with struct for
+  if (offload->task_type != offload->range_for)
+    return;
+
+  // Gather all atomic adds/subs destinations
+  // We use std::vector instead of std::set to keep an deterministic order here.
+  std::vector<GlobalPtrStmt *> atomic_destinations;
+  // TODO: this is again an abuse since it gathers nothing. Need to design a IR
+  // map/reduce system
+  auto linear_atomics =
+      irpass::analysis::gather_statements(offload, [&](Stmt *stmt) {
+        if (auto atomic_op = stmt->cast<AtomicOpStmt>()) {
+          if (is_atomic_op_linear(atomic_op->op_type)) {
+            // Local atomics does not count
+            if (auto dest = atomic_op->dest->cast<GlobalPtrStmt>()) {
+              if (std::find(atomic_destinations.begin(),
+                            atomic_destinations.end(),
+                            dest) == atomic_destinations.end()) {
+                atomic_destinations.push_back(dest);
+              }
+            }
+          }
+        }
+        return false;
+      });
+
+  std::vector<GlobalPtrStmt *> valid_reduction_values;
+
+  for (auto dest : atomic_destinations) {
+    // check if there is any other global load/store/atomic operations
+    auto related_global_mem_ops =
+        irpass::analysis::gather_statements(offload, [&](Stmt *stmt) {
+          if (auto load = stmt->cast<GlobalLoadStmt>()) {
+            if (maybe_same_address(load->ptr, dest)) {
+              return true;
+            }
+          } else if (auto store = stmt->cast<GlobalStoreStmt>()) {
+            if (maybe_same_address(store->ptr, dest)) {
+              return true;
+            }
+          } else if (auto atomic = stmt->cast<AtomicOpStmt>()) {
+            if (maybe_same_address(atomic->dest, dest)) {
+              return !is_atomic_op_linear(atomic->op_type);
+            }
+          }
+          for (auto &op : stmt->get_operands()) {
+            // Make sure the values of related atomic add operation are not
+            // used.
+            if (auto atomic = op->cast<AtomicOpStmt>()) {
+              if (maybe_same_address(atomic->dest, dest)) {
+                return true;
+              }
+            }
+          }
+          return false;  // Now we are sure the statement is not related to the
+                         // destination
+        });
+    TI_ASSERT(dest->width() == 1);
+    // We can only optimized reductions to global ptrs with form like loss[None]
+    // (0-D tensors) for now
+    if (related_global_mem_ops.empty() &&
+        dest->snodes[0]->type == SNodeType::place && dest->indices.empty()) {
+      valid_reduction_values.push_back(dest);
+    }
+  }
+
+  // We use 8 here (instead of sizeof(data_type) for TLS variable alignment
+  // (sizeof(f64) = 8).
+  constexpr std::size_t tls_stride = 8;
```

Comment:
I wonder whether we need to be more restrictive here. E.g.

```py
for i in x:
  loss[None] += x[i]
  ...
  foo(loss[None])  # is the reduction done at this point?
```

Reply:
Good point. This case should have been treated here: https://github.com/taichi-dev/taichi/pull/1296/files/32c37570cda915927f457f418dc524e979230481#diff-ab087800ac6701e94bc6e25c138a455dR59

When the return value of the atomic is used, we do not optimize.

Reply:
Thx!